ddtrace/tracer/writer.go

```diff
@@ -0,0 +1,277 @@
+	if math.IsInf(f, -1) {
+		return append(p, "-Infinity"...)
+	} else if math.IsInf(f, 1) {
+		return append(p, "Infinity"...)
+	}
+	abs := math.Abs(f)
+	if abs != 0 && (abs < 1e-6 || abs >= 1e21) {
+		p = strconv.AppendFloat(p, f, 'e', -1, 64)
+		// clean up e-09 to e-9
+		n := len(p)
+		if n >= 4 && p[n-4] == 'e' && p[n-3] == '-' && p[n-2] == '0' {
+			p[n-2] = p[n-1]
+			p = p[:n-1]
+		}
+	} else {
+		p = strconv.AppendFloat(p, f, 'f', -1, 64)
+	}
+	return p
+}
+
+func (h *logTraceWriter) encodeSpan(s *span) {
+	var scratch [maxFloatLength]byte
+	h.buf.WriteString(`{"trace_id":"`)
+	h.buf.Write(strconv.AppendUint(scratch[:0], uint64(s.TraceID), 16))
+	h.buf.WriteString(`","span_id":"`)
+	h.buf.Write(strconv.AppendUint(scratch[:0], uint64(s.SpanID), 16))
+	h.buf.WriteString(`","parent_id":"`)
+	h.buf.Write(strconv.AppendUint(scratch[:0], uint64(s.ParentID), 16))
+	h.buf.WriteString(`","name":"`)
+	h.buf.WriteString(s.Name)
+	h.buf.WriteString(`","resource":"`)
+	h.buf.WriteString(s.Resource)
+	h.buf.WriteString(`","error":`)
+	h.buf.Write(strconv.AppendInt(scratch[:0], int64(s.Error), 10))
+	h.buf.WriteString(`,"meta":{`)
+	first := true
+	for k, v := range s.Meta {
+		if first {
+			h.buf.WriteByte('"')
+			first = false
+		} else {
+			h.buf.WriteString(`,"`)
+		}
+		h.buf.WriteString(k)
+		h.buf.WriteString(`":"`)
+		h.buf.WriteString(v)
+		h.buf.WriteString(`"`)
+	}
+	h.buf.WriteString(`},"metrics":{`)
+	first = true
+	for k, v := range s.Metrics {
+		if first {
+			h.buf.WriteByte('"')
+			first = false
+		} else {
+			h.buf.WriteString(`,"`)
+		}
+		h.buf.WriteString(k)
+		h.buf.WriteString(`":`)
+		h.buf.Write(encodeFloat(scratch[:0], v))
+	}
+	h.buf.WriteString(`},"start":`)
+	h.buf.Write(strconv.AppendInt(scratch[:0], s.Start, 10))
+	h.buf.WriteString(`,"duration":`)
+	h.buf.Write(strconv.AppendInt(scratch[:0], s.Duration, 10))
+	h.buf.WriteString(`,"service":"`)
+	h.buf.WriteString(s.Service)
+	h.buf.WriteString(`"}`)
+}
+
+type encodingError struct {
+	cause      error
+	dropReason string
+}
+
+func (h *logTraceWriter) writeTrace(trace []*span) (n int, err *encodingError) {
+	startn := h.buf.Len()
+	if !h.hasTraces {
+		h.buf.WriteByte('[')
+	} else {
+		h.buf.WriteString(", [")
+	}
+	written := 0
+	for i, s := range trace {
+		n := h.buf.Len()
+		if i > 0 {
+			h.buf.WriteByte(',')
+		}
+		h.encodeSpan(s)
+		if h.buf.Len() > logPayloadLimit-len(logPayloadSuffix) {
+			if i == 0 {
+				h.buf.Truncate(startn)
+				if !h.hasTraces {
+					// This is the first span of the first trace, and it's too big.
+					return 0, &encodingError{cause: errors.New("span too large for payload"), dropReason: "trace_too_large"}
+				}
+				return 0, nil
+			}
+			h.buf.Truncate(n)
+			break
+		}
+		written++
+	}
+	h.buf.WriteByte(']')
+	h.hasTraces = true
+	return written, nil
+}
+
+func (h *logTraceWriter) add(trace []*span) {
+	for len(trace) > 0 {
+		n, err := h.writeTrace(trace)
+		if err != nil {
+			log.Error("lost a trace: %s", err.cause)
```

Comment:
```suggestion
			log.Error("Lost a trace: %s", err.cause)
```
I know we do this in the agent, not sure if it's a thing in the tracer too. Feel free to ignore if not.